onnxruntime/core/optimizer/matmul_add_fusion.cc

```diff
@@ -92,31 +153,58 @@ Status MatMulAddFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
-                                    "Gemm",
-                                    "fused Matmul and Add " + add_node.OpType(),
-                                    gemm_input_defs,
-                                    {});
+    auto gemm_output_defs = add_node.MutableOutputDefs();
+    if (need_reshape) {
+      auto add_reshape = [&](const std::vector<int64_t>& shape, Graph& graph, bool is_input) {
+        const std::string name = is_input ? "gemm_input" : "gemm_output";
+        ONNX_NAMESPACE::TensorProto shape_initializer_proto;
+        shape_initializer_proto.set_name(graph.GenerateNodeName(name + "_shape"));
+        shape_initializer_proto.add_dims(static_cast<int64_t>(shape.size()));
+        shape_initializer_proto.set_data_type(ONNX_NAMESPACE::TensorProto_DataType_INT64);
+        shape_initializer_proto.set_raw_data(shape.data(), shape.size() * sizeof(int64_t));
+        NodeArg* shape_arg = &graph_utils::AddInitializer(graph, shape_initializer_proto);
+        ONNX_NAMESPACE::TypeProto new_arg_type;
+        const ONNX_NAMESPACE::TensorProto_DataType element_type = static_cast<ONNX_NAMESPACE::TensorProto_DataType>(
+            gemm_input_defs[0]->TypeAsProto()->tensor_type().elem_type());
+        new_arg_type.mutable_tensor_type()->set_elem_type(element_type);
+        new_arg_type.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_value(m);
+        new_arg_type.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_value(is_input ? k : n);
+        NodeArg* new_arg = &graph.GetOrCreateNodeArg(graph.GenerateNodeArgName(name + "_reshape_arg"), &new_arg_type);
+        Node& reshape_node = graph.AddNode(graph.GenerateNodeName(name + "_reshape"), "Reshape", "Reshape for " + name,
+                                           {is_input ? gemm_input_defs[0] : new_arg, shape_arg},
+                                           {is_input ? new_arg : gemm_output_defs[0]});
+        reshape_node.SetExecutionProviderType(matmul_node.GetExecutionProviderType());
+        return new_arg;
+      };
+
+      gemm_input_defs[0] = add_reshape({m, k}, graph, true);
+      shape_values.back() = n;
+      gemm_output_defs[0] = add_reshape(shape_values, graph, false);
+    }
+
+    Node& gemm_node = graph.AddNode(graph.GenerateNodeName(matmul_node.Name() + "/MatMulAddFusion/"), "Gemm",
+                                    "fused Matmul and Add", gemm_input_defs, gemm_output_defs);
 
     // Assign provider to this new node. Provider should be same as the provider for old node.
     gemm_node.SetExecutionProviderType(matmul_node.GetExecutionProviderType());
 
-    // move output definitions and edges from act_node to gemm_node. delete gemm_node and act_node.
-    graph_utils::FinalizeNodeFusion(graph, {matmul_node, add_node}, gemm_node);
+    graph_utils::RemoveNodeOutputEdges(graph, matmul_node);
+    graph.RemoveNode(matmul_node.Index());
+    graph_utils::RemoveNodeOutputEdges(graph, add_node);
+    graph.RemoveNode(add_node.Index());
```

Comment:
FinalizeNodeFusion moved edges to the replacement node but we don't seem to be doing that anymore. Keeping the graph valid by handling the edges instead of requiring Graph::Resolve to fix them is generally preferred.